libfreerdp/core/nla.c

```diff
@@ -1748,82 +2086,95 @@ static BOOL nla_read_ts_credentials(rdpNla* nla, SecBuffer* data, size_t offset)
-		PasswordLength = nla->identity->PasswordLength;
-	}
+	TSCredentials_t cr = { 0 };
+	rdpSettings* settings = nla->settings;
 
-	if (nla->settings->DisableCredentialsDelegation && nla->identity)
+	if (settings->SmartcardLogon)
 	{
-		/* TSPasswordCreds */
-		nla->identity->DomainLength = 0;
-		nla->identity->UserLength = 0;
-		nla->identity->PasswordLength = 0;
-	}
+		TSSmartCardCreds_t smartcardCreds = { 0 };
+		TSCspDataDetail_t cspData = { 0 };
 
-	length = ber_sizeof_sequence(nla_sizeof_ts_credentials(nla));
+		smartcardCreds.pin = settings->SmartcardPin ? settings->SmartcardPin : "";
 
-	if (!nla_sec_buffer_alloc(&nla->tsCredentials, length))
-	{
-		WLog_ERR(TAG, "sspi_SecBufferAlloc failed!");
-		return FALSE;
-	}
+		/*smartcardCreds.userHint = settings->UserHint;
+		smartcardCreds.domainHint = settings->DomainHint;*/
+		smartcardCreds.cspData = &cspData;
 
-	s = Stream_New((BYTE*)nla->tsCredentials.pvBuffer, length);
+		cspData.keySpec = settings->KeySpec;
+		cspData.cspName = settings->CspName;
+		cspData.readerName = settings->ReaderName;
+		cspData.cardName = settings->CardName;
+		cspData.containerName = settings->ContainerName;
 
-	if (!s)
+		length = ber_sizeof_nla_TSSmartCardCreds(&smartcardCreds);
+		credsContentStream = Stream_New(NULL, length);
+		if (!credsContentStream)
+			return FALSE;
+
+		if (ber_write_nla_TSSmartCardCreds(credsContentStream, &smartcardCreds) == 0)
+			return FALSE;
+
+		cr.credType = 2;
+	}
+	else
 	{
-		sspi_SecBufferFree(&nla->tsCredentials);
-		WLog_ERR(TAG, "Stream_New failed!");
-		return FALSE;
+		TSPasswordCreds_t passCreds = { 0 };
+
+		if (!nla->settings->DisableCredentialsDelegation && nla->identity)
+		{
+			passCreds.userNameLen = nla->identity->UserLength * 2;
+			passCreds.userName = (BYTE*)nla->identity->User;
+
+			passCreds.domainNameLen = nla->identity->DomainLength * 2;
+			passCreds.domainName = (BYTE*)nla->identity->Domain;
+
+			passCreds.passwordLen = nla->identity->PasswordLength * 2;
+			passCreds.password = (BYTE*)nla->identity->Password;
+		}
+
+		length = ber_sizeof_nla_TSPasswordCreds(&passCreds);
+		credsContentStream = Stream_New(NULL, length);
+		if (!credsContentStream)
+			return FALSE;
+
+		ber_write_nla_TSPasswordCreds(credsContentStream, &passCreds);
+
+		cr.credType = 1;
 	}
 
-	nla_write_ts_credentials(nla, s);
+	cr.credentialsLen = length;
+	cr.credentials = Stream_Buffer(credsContentStream);
 
-	if (nla->settings->DisableCredentialsDelegation && nla->identity)
+	length = ber_sizeof_nla_TSCredentials(&cr);
+	if (!nla_sec_buffer_alloc(&nla->tsCredentials, length))
 	{
-		/* TSPasswordCreds */
-		nla->identity->DomainLength = DomainLength;
-		nla->identity->UserLength = UserLength;
-		nla->identity->PasswordLength = PasswordLength;
+		WLog_ERR(TAG, "sspi_SecBufferAlloc failed!");
+		goto out;
 	}
 
-	Stream_Free(s, FALSE);
-	return TRUE;
+	s = Stream_StaticInit(&staticRetStream, (BYTE*)nla->tsCredentials.pvBuffer, length);
+	ber_write_nla_TSCredentials(s, &cr);
```

Comment:
missing length check?

Reply:
I don't see any pertinent check, we allocate the correct size a few line before, and `Stream_StaticInit` never fails...